evm/src/invariant_fuzz.rs

```diff
@@ -0,0 +1,363 @@
+        })
+    }
+}
+
+/// The outcome of an invariant fuzz test
+pub struct InvariantFuzzTestResult {
+    pub invariants: BTreeMap<String, Option<InvariantFuzzError>>,
+    /// Every successful fuzz test case
+    pub cases: FuzzedCases,
+}
+
+pub struct InvariantFuzzError {
+    /// The proptest error occurred as a result of a test case
+    pub test_error: TestError<Vec<(Address, Bytes)>>,
+    /// The return reason of the offending call
+    pub return_reason: Reason,
+    /// The revert string of the offending call
+    pub revert_reason: String,
+    /// Address of the invariant asserter
+    pub addr: Address,
+    /// Function data for invariant check
+    pub func: ethers::prelude::Bytes,
+}
+
+/// Container type for all successful test cases
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(transparent)]
+pub struct InvariantFuzzedCases {
+    cases: Vec<FuzzCase>,
+}
+
+impl InvariantFuzzedCases {
+    pub fn new(mut cases: Vec<FuzzCase>) -> Self {
+        cases.sort_by_key(|c| c.gas);
+        Self { cases }
+    }
+
+    pub fn cases(&self) -> &[FuzzCase] {
+        &self.cases
+    }
+
+    pub fn into_cases(self) -> Vec<FuzzCase> {
+        self.cases
+    }
+
+    /// Returns the median gas of all test cases
+    pub fn median_gas(&self) -> u64 {
+        let mid = self.cases.len() / 2;
+        self.cases.get(mid).map(|c| c.gas).unwrap_or_default()
+    }
+
+    /// Returns the average gas use of all test cases
+    pub fn mean_gas(&self) -> u64 {
+        if self.cases.is_empty() {
+            return 0
+        }
+
+        (self.cases.iter().map(|c| c.gas as u128).sum::<u128>() / self.cases.len() as u128) as u64
+    }
+
+    pub fn highest(&self) -> Option<&FuzzCase> {
+        self.cases.last()
+    }
+
+    pub fn lowest(&self) -> Option<&FuzzCase> {
+        self.cases.first()
+    }
+
+    pub fn highest_gas(&self) -> u64 {
+        self.highest().map(|c| c.gas).unwrap_or_default()
+    }
+
+    pub fn lowest_gas(&self) -> u64 {
+        self.lowest().map(|c| c.gas).unwrap_or_default()
+    }
+}
+
+/// Data of a single fuzz test case
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct InvariantFuzzCase {
+    /// The calldata used for this fuzz test
+    pub calldata: Bytes,
+    // Consumed gas
+    pub gas: u64,
+}
+
+pub fn invariant_strat(
+    depth: usize,
+    contracts: BTreeMap<Address, (String, Abi)>,
+) -> BoxedStrategy<Vec<(Address, Bytes)>> {
+    let iters = 1..depth + 1;
+    proptest::collection::vec(gen_call(contracts), iters).boxed()
+}
+
+fn gen_call(contracts: BTreeMap<Address, (String, Abi)>) -> BoxedStrategy<(Address, Bytes)> {
+    let random_contract = select_random_contract(contracts);
+    random_contract
+        .prop_flat_map(move |(contract, abi)| {
+            let func = select_random_function(abi);
+            func.prop_flat_map(move |func| fuzz_calldata(contract, func))
+        })
+        .boxed()
+}
+
+fn select_random_contract(
```

Comment:
> it will invoke methods from any non-testing contract available after the setUp() function has been run, checking the invariant* after each run.
Unless I misunderstood something, this is what is happening right now. Any contract **created** during the `setUp` function is used.

>all addresses returned by a call to targetContracts()
As a way to filter-out contracts we don't want to fuzz? Because if it includes addresses from a forked state, then it requires the use of etherscan

Reply:
oh nice